### PR TITLE
Annotations - prevent crop-box resizing against borders

### LINKED ISSUE #4007: [Annotations] Drag to Move Crop Selection Resizes Box Against Borders
#### STR

1. Start cropping tool
1. Drag to select an area
1. Drag the entire selection box towards the left or right border of the shot

#### Result

The selection box is resized as the user continues to drag against the border.

#### Expected

The selection box simply stop moving once it's aligned to the border.

### PR DESCRIPTION
**Pull request by niharikak101**
Fix #4007

### CODE REVIEW

File: server/src/pages/shot/editor.js [hunk 1]
```diff
@@ -270,14 +270,22 @@ exports.Editor = class Editor extends React.Component {
...
         for (let moveDir of moveX) {
-          selectedPos[moveDir] = this.truncateX(resizeStartSelected[moveDir] + diffX);
+          if (resizeDirection == "move" && (selectedPos.left == 0 && resizeStartSelected.left + diffX < 0 || selectedPos.right == this.canvasWidth && resizeStartSelected.right + diffX > this.canvasWidth)) {
+            break;
+          } else {
+            selectedPos[moveDir] = this.truncateX(resizeStartSelected[moveDir] + diffX);
+          }
         }
       }
       if (movement[1]) {
```
Comment (chenba) [inline server/src/pages/shot/editor.js]: I just noticed this `if` is indented one more level than the above `if`, which is a little confusing at a glace.

File: server/src/pages/shot/editor.js [hunk 2]
```diff
@@ -270,14 +270,22 @@ exports.Editor = class Editor extends React.Component {
         let moveX = movement[0];
         moveX = moveX == "*" ? ["x1", "x2"] : [moveX];
         for (let moveDir of moveX) {
-          selectedPos[moveDir] = this.truncateX(resizeStartSelected[moveDir] + diffX);
+          if (resizeDirection == "move" && (selectedPos.left == 0 && resizeStartSelected.left + diffX < 0 || selectedPos.right == this.canvasWidth && resizeStartSelected.right + diffX > this.canvasWidth)) {
```
Comment (chenba) [inline server/src/pages/shot/editor.js]: This line is long and difficult to read.  Perhaps assign the conditions into variables so they are named, so it reads `if (someThing && someOtherThing && oneMoreThing)`.

Also, this `if` does not need to be inside the `for` loop.

**Comment by chenba** [comment]
I found that if I drag to move the selection box very quickly back and forth between the left and right edges of the image, I can get the selection box to resize.

**Comment by niharikak101** [comment]
I added a function to preserve the original width and height of the selection box if the user hits the borders!

File: server/src/pages/shot/editor.js [hunk 3]
```diff
@@ -262,27 +262,53 @@ exports.Editor = class Editor extends React.Component {
...
-        let moveX = movement[0];
-        moveX = moveX == "*" ? ["x1", "x2"] : [moveX];
-        for (let moveDir of moveX) {
-          selectedPos[moveDir] = this.truncateX(resizeStartSelected[moveDir] + diffX);
-        }
+    let checkLeftBorder = selectedPos.left == 0 && resizeStartSelected.left + diffX <= 0;
+    let checkRightBorder = selectedPos.right == this.canvasWidth && resizeStartSelected.right + diffX >= this.canvasWidth;
+    let checkTopBorder = selectedPos.top == 0 && resizeStartSelected.top + diffY <= 0;
+    let checkBottomBorder = selectedPos.bottom == this.canvasHeight && resizeStartSelected.bottom + diffY >= this.canvasHeight;
+    let checkResizeDirection = resizeDirection == "move";
```
Comment (chenba) [inline server/src/pages/shot/editor.js]: Nit: perhaps name these variables in a `isCondition` convention.  That'd read more naturally.

**Pull request merged**